### DISCUSSION TITLE
Does Fastify provide complete encapsulation for all runtime errors, including those thrown by it's dependent modules?

### DISCUSSION THREAD

**Discussion by U-4-E-A** [fastify/fastify#5054] [Q&A] [answered]
****EDIT**: it would seem express _can_ to this type of error handling after all... it looks like I didn't configure my error handler correctly. Still going to look into Fastify though.**

I am currently using `express` but I am considering moving to another HTTP server due to an issue I have with how `express` handles errors that are thrown by modules in middleware. The example I have of this is from `express-session`: -

```
    app.use(
      expressSession({
        [...]
        secret: process.env.SESSIONS_SECRET
      })
    )
```

If `process.env.SESSIONS_SECRET` is null then this throws an error when express handles a request but NOT during initialisation of the server - the server will start OK but then crash when processing a request. This sends back a 500 and horrible stack trace, which is visible in the browser's network tab. This is something I wish to avoid and would like graceful error handling, which it seems `express` doesn't seem to provide in this context.

To clarify 2 points: -

Firstly, it seems no error handler in `express` will catch this error as `express-session` throwing the error simply crashes `express` and it is not caught. `express` only seems to catch errors from within it's own core code rather than that of it's dependent modules.

Secondly, I use types everywhere possible. `process.env` variables can be `string` or `null`. So, in this case, the only option is to use `ts-ignore` to stop TS complaining if I want to use `process.env`. This is not desirable and it does not address the bigger issue that when dependent modules throw errors then `express` simply doesn't seem to have a mechanism of catching them. I want to be assured that if my server throws a 500 then I can stop the server from returning the stack trace.

I have discussed the issue with `express` on reddit and people don't seem to understand that: -

1. The error cannot (it seems) be caught by `express`
2. That there is the larger issue of being unable to control exposure of the stack trace

Does Fastify have the functionality to intercept errors before the HTTP response is sent to control what is returned by it regardless of where the error is thrown? Essentially, a global try->catch wrapper for the HTTP server.

Thanks.

**Comment by jsumners** [comment]
Have you considered writing a minimal implementation to test for your requirments?

**Reply by U-4-E-A** [reply]
A minimal implementation of Fastify to mimic the error and see if it is caught?

**Reply by jsumners** [reply]
A minimal implementation of whatever application you want to write with Fastify.

**Reply by U-4-E-A** [reply]
I will try writing a test app to see what happens with this type of error. I was wondering if Fastify's devs were aware of this specific issue (it is understandable that they wouldn't be, it's definitely an edge case) before attempting this. I will write a minimal implementation as you suggest and see what happens. Thanks.

**Comment by Eomm** [comment, marked as answer]
> Does Fastify have the functionality to intercept errors before the HTTP response is sent to control what is returned by it regardless of where the error is thrown? Essentially, a global try->catch wrapper for the HTTP server.

Yes, checkout this old but still gold article:
https://backend.cafe/fastify-error-handlers
